apis/Google.Cloud.Iam.V1/Google.Cloud.Iam.V1/PolicyPartial.cs

```diff
@@ -26,14 +26,20 @@ public partial class Policy
         /// <summary>
         /// Adds the specified member to the specified role. If the role does
         /// not already exist, it is created.
+        /// This method will fail with an <see cref="InvalidOperationException"/>
+        /// if it is called on a Policy with a <see cref="Version"/> greater than 1,
+        /// as that indicates a more complicated policy than this method is prepared
+        /// to handle. Changes to such policies must be made manually.
         /// </summary>
         /// <param name="role">The role to add the member to. Must not be null or empty.</param>
         /// <param name="member">The member to add to the role. Must not be null or empty.</param>
         /// <returns><c>true</c> if the policy was changed; <c>false</c> if the member already existed in the role.</returns>
+        /// <exception cref="InvalidOperationException">The <see cref="Version"/> is greater than 1.</exception>
```

Comment:
From my reading, this also only happens if SupportedIamPolicyVersion is set to 3, right?
That is, without any code modification (SupportedIamPolicyVersion=3), existing programs will not get an exception *here*, but when they *fetch* a policy with Version=3 and SupportedIamPolicyVersion=default.

Though I think it's probably safer to make the helpers invalid/throw exception when SupportedIamPolicyVersion=3

Reply:
Yes, this shouldn't happen unless you request a V3 policy explicitly.

But:

- You could request a V3 policy explicitly without realizing that it effectively invalidates the use of these methods
- There could be multiple teams in the same company, where one team passes a policy that it's fetched with v3 support to code written by a team that hasn't even heard of conditions

So as you say, it's safer to throw here as well. Very much by design.

Reply:
I think it's also possible to:

* Request with SupportedIamPolicyVersion=3
* Policy returned by server is Version=1
* Modify bindings (e.g., add a binding with a condition)
* Use the v1-only helpers

No exception is thrown because the version in the policy is only set to 3 in ModifyRequest

Yes, quite an edge case.

Reply:
> But would be prevented by setting Policy=3 in GetIamPolicy rather than SetIamPolicy.

I don't see how. The scenario you described *does* put Policy=3 in GetIamPolicy. Or are you saying that we should modify the Policy object returned by the server to be v3 even if it doesn't contain any conditions? If that's the right thing to do, the server should do it. (And we should persuade the IAM team of that.)

I feel I may have misunderstood what you're suggesting though. Could you clarify the difference between what's in this PR and what you'd like to be in there?

Reply:
> Or are you saying that we should modify the Policy object returned by the server to be v3 even if it doesn't contain any conditions?

Yes, that's what I was saying. I meant "PolicyVersion=3" (PolicyVersion being different to RequestedPolicyVersion).

> If that's the right thing to do, the server should do it. (And we should persuade the IAM team of that.)

Maybe. If the underlying bindings were not directly editable, then it would be a valid signal to throw a runtime error. Here's a valid way it could be used (pseudocode):
```
policy = GetIamPolicy(RequestedPolicyVersion=3)

assert(policy.PolicyVersion == 1) // there are no conditions in this policy
policy.AddMember("user:foo@foo.com", "reader") // a "V1" helper

policy.Raw()[0].Condition = ...

policy.AddMember("user:foo@foo.com", "reader") // throws an exception because the call to `Raw()` above means the policy may not be compatible with V1 helpers.
```

Does that make sense? I know we're deep in the weeds with corner cases and hypotheticals, just trying to make sure we are on the safe/conservative side.

Reply:
I don't think that's feasible, really. The bindings are already directly visible - we can't start breaking that now.
I'd also say that for better or worse, the way the API is structured discourages this.

Now we could make the helper methods detect the presence of a condition in any binding, of course - that would be conservative, but it might also be not-conservative-enough if (and I hope this doesn't happen) another v3-only aspect is introduced as well.

Fundamentally all of this feels fragile, but I think using just the version is at least a simple-to-understand approach.